src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp

```diff
@@ -172,69 +173,47 @@ bool AsynchronousReadIndirectBufferFromRemoteFS::nextImpl()
 
     Stopwatch watch;
     CurrentMetrics::Increment metric_increment{CurrentMetrics::AsynchronousReadWait};
-
-    size_t size = 0;
-    size_t bytes_read = 0;
+    size_t size, offset;
 
     if (prefetch_future.valid())
     {
-        ProfileEvents::increment(ProfileEvents::RemoteFSPrefetchedReads);
-
-        size_t offset = 0;
-        {
-            auto result = prefetch_future.get();
-            size = result.size;
-            offset = result.offset;
-            LOG_TEST(log, "Current size: {}, offset: {}", size, offset);
-
-            /// If prefetch_future is valid, size should always be greater than zero.
-            assert(offset <= size);
-            bytes_read = size - offset;
-
-            ProfileEvents::increment(ProfileEvents::AsynchronousReadWaitMicroseconds, watch.elapsedMicroseconds());
-        }
+        ElapsedUSProfileEventIncrement measure_time(ProfileEvents::AsynchronousRemoteReadWaitMicroseconds);
 
+        std::tie(size, offset) = prefetch_future.get();
+        prefetch_future = {};
         prefetch_buffer.swap(memory);
 
-        /// Adjust the working buffer so that it ignores `offset` bytes.
-        internal_buffer = Buffer(memory.data(), memory.data() + memory.size());
-        working_buffer = Buffer(memory.data() + offset, memory.data() + size);
-        pos = working_buffer.begin();
+        ProfileEvents::increment(ProfileEvents::RemoteFSPrefetchedReads);
     }
     else
     {
-        ProfileEvents::increment(ProfileEvents::RemoteFSUnprefetchedReads);
+        ElapsedUSProfileEventIncrement measure_time(ProfileEvents::AsynchronousRemoteReadWaitMicroseconds);
 
         chassert(memory.size() == read_settings.remote_fs_buffer_size);
-        auto result = asyncReadInto(memory.data(), memory.size()).get();
-        size = result.size;
-        auto offset = result.offset;
-
-        LOG_TEST(log, "Current size: {}, offset: {}", size, offset);
-
-        assert(offset <= size);
-        bytes_read = size - offset;
+        std::tie(size, offset) = impl->readInto(memory.data(), memory.size(), file_offset_of_buffer_end, bytes_to_ignore);
```

Comment:
The point is removing read with threadpool if is is synchronous read (not from prefetch)

test:

test_s3 and test_local are tables with the same data but the former is on s3 and the second is local

from local table

```
ip-172-31-4-213.eu-west-1.compute.internal :) SELECT count() FROM test_local WHERE s IN (SELECT toString(arrayJoin(range(1000)) * 100000 + 55555)) SETTINGS enable_filesystem_cache=1

1 row in set. Elapsed: 0.060 sec. Processed 15.90 million rows, 260.63 MB (266.41 million rows/s., 4.37 GB/s.)
```

from cache over s3 before (in many runs result is in range **0.8-1**):

```
ip-172-31-4-213.eu-west-1.compute.internal :) SELECT count() FROM test_s3 WHERE s IN (SELECT toString(arrayJoin(range(1000)) * 100000 + 55555)) SETTINGS enable_filesystem_cache=1

1 row in set. Elapsed: 0.091 sec. Processed 15.88 million rows, 260.41 MB (174.51 million rows/s., 2.86 GB/s.)
```

from cache over s3 after  (in many runs result in range **0.65-0.75**):

```
ip-172-31-4-213.eu-west-1.compute.internal :) SELECT count() FROM test_s3 WHERE s IN (SELECT toString(arrayJoin(range(1000)) * 100000 + 55555)) SETTINGS enable_filesystem_cache=1

1 row in set. Elapsed: 0.068 sec. Processed 15.88 million rows, 260.41 MB (233.56 million rows/s., 3.83 GB/s.)
```